ksqldb-engine/src/main/java/io/confluent/ksql/engine/QueryMonitor.java

```diff
@@ -0,0 +1,237 @@
+  private final Map<QueryId, RetryEvent> queriesRetries = new HashMap<>();
+
+  private volatile boolean closed = false;
+
+  public QueryMonitor(final KsqlConfig ksqlConfig, final KsqlEngine ksqlEngine) {
+    this(
+        ksqlEngine,
+        Executors.newSingleThreadExecutor(r -> new Thread(r, QueryMonitor.class.getName())),
+        ksqlConfig.getLong(KsqlConfig.KSQL_QUERY_RETRY_BACKOFF_INITIAL_MS),
+        ksqlConfig.getLong(KsqlConfig.KSQL_QUERY_RETRY_BACKOFF_MAX_MS),
+        CURRENT_TIME_MILLIS_TICKER
+    );
+  }
+
+  @VisibleForTesting
+  QueryMonitor(
+      final KsqlEngine ksqlEngine,
+      final ExecutorService executor,
+      final long retryBackoffInitialMs,
+      final long retryBackoffMaxMs,
+      final Ticker ticker
+  ) {
+    this.retryBackoffInitialMs = retryBackoffInitialMs;
+    this.retryBackoffMaxMs = retryBackoffMaxMs;
+    this.ksqlEngine = ksqlEngine;
+    this.executor = executor;
+    this.ticker = ticker;
+  }
+
+  public void start() {
+    executor.execute(new Runner());
+    executor.shutdown();
+  }
+
+  @Override
+  public void close() {
+    closed = true;
+
+    try {
+      executor.awaitTermination(SHUTDOWN_TIMEOUT_MS, TimeUnit.MILLISECONDS);
+    } catch (final InterruptedException e) {
+      Thread.currentThread().interrupt();
+    }
+  }
+
+  boolean isClosed() {
+    return closed;
+  }
+
+  void restartFailedQueries() {
+    // Collect a list of new queries in ERROR state
+    ksqlEngine.getPersistentQueries().stream()
+        .filter(QueryMetadata::isError)
+        .filter(query -> !queriesRetries.containsKey(query.getQueryId()))
+        .map(QueryMetadata::getQueryId)
+        .forEach(queryId -> queriesRetries.put(queryId, newRetryEvent(queryId)));
+
+    maybeRestartQueries();
+  }
+
+  private void maybeRestartQueries() {
+    final long now = ticker.read();
+
+    // Restart queries that has passed the waiting timeout
+    final List<QueryId> deleteRetryEvents = new ArrayList<>();
+    queriesRetries.entrySet().stream()
+        .forEach(mapEntry -> {
+          final QueryId queryId = mapEntry.getKey();
+          final RetryEvent retryEvent = mapEntry.getValue();
+          final Optional<PersistentQueryMetadata> query = ksqlEngine.getPersistentQuery(queryId);
+
+          // Query was terminated manually if no present
+          if (!query.isPresent()) {
+            deleteRetryEvents.add(queryId);
+          } else if (query.get().isError() && now > retryEvent.nextRestartTimeMs()) {
+            // Retry again if it's still in ERROR state
+            retryEvent.restart();
+          } else if (now > retryEvent.queryHealthyTime()) {
+            // Clean the errors queue & delete the query from future retries now the query is
+            // healthy
+            query.ifPresent(QueryMetadata::clearErrors);
+            deleteRetryEvents.add(queryId);
+          }
+        });
+
+    deleteRetryEvents.stream().forEach(queriesRetries::remove);
```

Comment:
might be good to add a summary log line here (e.g. `restarted queries [X, Y], marked queries [Z] as healthy, could not find queries [A]`)